src/Common/MatchGenerator.cpp

```diff
@@ -0,0 +1,514 @@
+                chassert(consumed <= size);
+                out += consumed;
+                size -= consumed;
+                total_size += consumed;
+            }
+            return total_size;
+        }
+
+        size_t getRequiredSize() override
+        {
+            return max_repeat * func->getRequiredSize();
+        }
+
+    private:
+        NodeFunctionPtr func;
+        int min_repeat = 0;
+        int max_repeat = 0;
+    };
+
+    class RegexpCharClassFunction : public NodeFunction
+    {
+        using CharRanges = std::vector<std::pair<re2::Rune, re2::Rune>>;
+
+    public:
+        explicit RegexpCharClassFunction(Regexp * re_)
+        {
+            CharClass * cc = re_->cc();
+            chassert(cc);
+            if (cc->empty())
+                throw DB::Exception(DB::ErrorCodes::BAD_ARGUMENTS, "kRegexpCharClass is empty");
+
+            char_count = cc->size();
+            char_ranges.reserve(std::distance(cc->begin(), cc->end()));
+
+            for (const auto range: *cc)
+            {
+                char_ranges.emplace_back(range.lo, range.hi);
+            }
+        }
+
+        size_t operator () (char * out, size_t size) override
+        {
+            chassert(UTFmax <= size);
+
+            std::uniform_int_distribution<int> distribution(1, char_count);
+            int chosen = distribution(thread_local_rng);
+            int count_down = chosen;
+
+            auto it = char_ranges.begin();
+            for (; it != char_ranges.end(); ++it)
+            {
+                auto [lo, hi] = *it;
+                auto range_len = hi - lo + 1;
+                if (count_down <= range_len)
+                    break;
+                count_down -= range_len;
+            }
+
+            if (it == char_ranges.end())
+                throw DB::Exception(DB::ErrorCodes::LOGICAL_ERROR,
+                                    "Unable to choose the rune. Runes {}, ranges {}, chosen {}",
+                                    char_count, char_ranges.size(), chosen);
+
+            auto [lo, _] = *it;
+            Rune r = lo + count_down - 1;
+            int n = re2::runetochar(out, &r);
+            return n;
+        }
+
+        size_t getRequiredSize() override
+        {
+            return UTFmax;
+        }
+
+    private:
+        int char_count = 0;
+        CharRanges char_ranges;
+    };
+
+    class RegexpLiteralStringFunction : public NodeFunction
+    {
+    public:
+        explicit RegexpLiteralStringFunction(Regexp * re_)
+        {
+            if (re_->nrunes() == 0)
+                return;
+
+            char buffer[UTFmax];
+            for (int i = 0; i < re_->nrunes(); ++i)
+            {
+                int n = re2::runetochar(buffer, &re_->runes()[i]);
+                literal_string += String(buffer, n);
+            }
+        }
+
+        size_t operator () (char * out, size_t size) override
+        {
+            chassert(literal_string.size() <= size);
```

Comment:
As you see, there are allocation optimized when result is generated (not when generator is built), otherwise generation is slow. Now there is exactly one allocation with result string. That is possible due to calculation max size of the result string when generator is buld, see `getRequiredSize`.

That `chassert(literal_string.size() <= size)` checks that there is still enough space in `char * out` to hold result of the  current gerators node.  In other places such checks look like `chassert(consumed <= size);`. But for the leafs of the tree, we know what we exactly going to write to `char * out`. Like `chassert(UTFmax <= size);`